### PR TITLE
feat: Refresh access token on appeal submission - part 1

### PR DESCRIPTION
**Pull request by n-r-b**
### JIRA link
https://companieshouse.atlassian.net/browse/LFA-1704 (Part 1)

### Change description
* Created refresh token service
* Created refresh token intercepter which performs token refresh when appeal submission/retrieval request has failed due to a 401 unauthorised response code, and retries the original request
* Updated AppealService tests to cover refresh token scenario

* TO-DO (Part 2): Update access token in session

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes look good on mobile
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are commited to keeping commit history clean, consistent and linear. To achive this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/Constants.ts [hunk 1]
```diff
@@ -1 +1,3 @@
 export const APP_NAME = 'lfp-appeals-frontend';
+
+export const GRANT_TYPE = 'refresh_token';
```
Comment (akaJSaunders) [inline src/Constants.ts:3]: The name of the variable seems rather vague, especially without the context of this PR

File: src/Constants.ts [hunk 2]
```diff
@@ -1 +1,3 @@
 export const APP_NAME = 'lfp-appeals-frontend';
+
```
Comment (akaJSaunders) [inline src/Constants.ts:2]: Don't think this empty line is necessary

File: src/modules/appeals-service/AppealsService.ts [hunk 3]
```diff
@@ -1,45 +1,55 @@
...
     }
 
-    public async getAppeal(companyNumber: string, appealId: string, token: string): Promise<Appeal> {
+    public async getAppeal(companyNumber: string, appealId: string, accessToken: string,
+                           refreshToken: string): Promise<Appeal> {
 
         this.checkArgumentOrThrow(companyNumber, 'Company number is missing');
         this.checkArgumentOrThrow(appealId, 'Appeal id is missing');
-        this.checkArgumentOrThrow(token, 'Token is missing');
+        this.checkArgumentOrThrow(accessToken, 'Token is missing');
```
Comment (akaJSaunders) [inline src/modules/appeals-service/AppealsService.ts:24]: Now that we have a refresh token, maybe its worth updating this message

```suggestion
        this.checkArgumentOrThrow(accessToken, 'Access token is missing');
```

File: src/modules/appeals-service/AppealsService.ts [hunk 4]
```diff
@@ -1,45 +1,55 @@
...
+
```
Comment (akaJSaunders) [inline src/modules/appeals-service/AppealsService.ts:25]: should we be checking for refresh token here aswell?: 
```suggestion
        this.checkArgumentOrThrow(refreshToken, 'Refresh token is missing');
```

File: src/modules/appeals-service/AppealsService.ts [hunk 5]
```diff
@@ -84,4 +83,38 @@ export class AppealsService {
...
+        this.axiosInstance.interceptors.response.use((response: AxiosResponse) => {
+            return response;
+        }, async (error: AxiosError) => {
+
+            const response: AxiosResponse | undefined = error.response;
+
+            if (response && response.status === UNAUTHORIZED) {
+                loggerInstance()
+                    .info(`${AppealsService.name} - create appeal failed with:  ${response.status} - attempting token refresh`);
```
Comment (akaJSaunders) [inline src/modules/appeals-service/AppealsService.ts:112]:
```suggestion
                    .info(`${AppealsService.name} - create appeal failed with: ${response.status} - attempting token refresh`);
```

File: src/modules/refresh-token-service/RefreshTokenService.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+        return await axios
+            .post(this.uri, requestParams)
+            .then((response: AxiosResponse<RefreshTokenData>) => {
+                if (response.status === OK && response.data) {
+                    loggerInstance()
+                        .debug(`${RefreshTokenService.name} - refresh: created new access token - ${response.data.access_token}`);
+                    return response.data.access_token;
+                }
+                throw new Error('Could not refresh token');
```
Comment (akaJSaunders) [inline src/modules/refresh-token-service/RefreshTokenService.ts:47]:
```suggestion
                throw new Error('Could not refresh access token');
```

File: test/controllers/processors/AppealStorageFormSubmissionProcessor.test.ts [hunk 7]
```diff
@@ -35,6 +35,7 @@ describe('AppealStorageForSubmissionProcessor', () => {
     };
 
     const token: string = 'abc';
```
Comment (akaJSaunders) [inline test/controllers/processors/AppealStorageFormSubmissionProcessor.test.ts:37]: we should probably refactor 'token' to 'accessToken'

File: test/modules/appeals-service/AppealsService.test.ts [hunk 8]
```diff
@@ -193,37 +235,92 @@ describe('AppealsService', () => {
...
 
             const returnedAppeal = await appealsService
-                .getAppeal(appeal.penaltyIdentifier.companyNumber, APPEAL_ID, BEARER_TOKEN);
+                .getAppeal(appeal.penaltyIdentifier.companyNumber, APPEAL_ID, BEARER_TOKEN, REFRESH_TOKEN);
 
             expect(returnedAppeal).to.deep.eq(appeal);
 
         });
 
+        it('should return an appeal when valid arguments are provided after refreshing an expired token', async () => {
```
Comment (akaJSaunders) [inline test/modules/appeals-service/AppealsService.test.ts:269]: I think its worth being as specific as possible
```suggestion
        it('should return an appeal when valid arguments are provided after refreshing an expired access token', async () => {
```

File: test/modules/refresh-token-service/RefreshTokenService.test.ts [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+                try {
+                    await refreshTokenService.refresh(ACCESS_TOKEN, refreshToken as any);
+                } catch (err) {
+                    expect(err).to.be.instanceOf(Error)
+                        .and.to.haveOwnProperty('message').equal('Refresh token is missing');
+                }
+            });
+        });
+
+        it('should refresh token', async () => {
```
Comment (akaJSaunders) [inline test/modules/refresh-token-service/RefreshTokenService.test.ts:56]:
```suggestion
        it('should refresh access token', async () => {
```

File: test/modules/refresh-token-service/RefreshTokenService.test.ts [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        });
+
+        it('should return status 400 when refresh token is invalid', async () => {
+
+            nock(HOST)
+                .post(URI)
+                .reply(BAD_REQUEST, refreshTokenDataInvalid);
+
+            try {
+                await refreshTokenService.refresh(ACCESS_TOKEN, REFRESH_TOKEN);
```
Comment (franciscog-kainos) [inline test/modules/refresh-token-service/RefreshTokenService.test.ts:89]: How do you know this test does not fail? If the refresh is successful the test will pass

File: test/modules/refresh-token-service/RefreshTokenService.test.ts [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+        });
+
+        it('should return status 500 when internal server error', async () => {
+
+            nock(HOST)
+                .post(URI)
+                .reply(INTERNAL_SERVER_ERROR);
+
+            try {
+                await refreshTokenService.refresh(ACCESS_TOKEN, REFRESH_TOKEN);
```
Comment (franciscog-kainos) [inline test/modules/refresh-token-service/RefreshTokenService.test.ts:103]: Same here

File: test/modules/refresh-token-service/RefreshTokenService.test.ts [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        });
+
+        it('should throw error when response data is empty', async () => {
+
+            nock(HOST)
+                .post(URI)
+                .reply(OK, {});
+
+            try {
+                await refreshTokenService.refresh(ACCESS_TOKEN, REFRESH_TOKEN);
```
Comment (franciscog-kainos) [inline test/modules/refresh-token-service/RefreshTokenService.test.ts:75]: How do you know this test does not fail? If the refresh is successful the test will pass

File: test/modules/refresh-token-service/RefreshTokenService.test.ts [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+                        .and.to.haveOwnProperty('message').equal('Access token is missing');
+                }
+            });
+        });
+
+        it('should throw an error when refresh token is not defined', () => {
+
+            [undefined, null].forEach(async refreshToken => {
+                try {
+                    await refreshTokenService.refresh(ACCESS_TOKEN, refreshToken as any);
```
Comment (franciscog-kainos) [inline test/modules/refresh-token-service/RefreshTokenService.test.ts:48]: If the refresh does not throw then the test will pass

File: test/modules/refresh-token-service/RefreshTokenService.test.ts [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        'error': 'unauthorized_client'
+    };
+
+    describe('Refreshing token', () => {
+
+        it('should throw an error when access token is not defined', () => {
+
+            [undefined, null].forEach(async accessToken => {
+                try {
+                    await refreshTokenService.refresh(accessToken as any, REFRESH_TOKEN);
```
Comment (franciscog-kainos) [inline test/modules/refresh-token-service/RefreshTokenService.test.ts:36]: Same as comments below

File: test/modules/refresh-token-service/RefreshTokenService.test.ts (hunk 13 above)
Comment (franciscog-kainos) [reply test/modules/refresh-token-service/RefreshTokenService.test.ts:48]: `assert.fail()`

**Pull request merged**